deviceModeInit.js

```diff
@@ -95,6 +96,117 @@ var rudderTracking = (function () {
       .catch((error) => {
         console.log("Error occurred while updating cart:", error);
       });
+
+    const productIsVisible = (window, element) => {
+      const viewportWidth = window.innerWidth;
+      const viewportHeight = window.innerHeight;
+      const { top, bottom, left, right, height } =
+        element.getBoundingClientRect();
+
+      if (
+        top < viewportHeight &&
+        bottom > 0 &&
+        left < viewportWidth &&
+        right > 0
+      ) {
+        let pixelsVisible = height;
+
+        if (top < 0) {
+          pixelsVisible += top;
+        }
+        if (bottom > viewportHeight) {
+          pixelsVisible += viewportHeight - bottom;
+        }
+        const percentVisible = pixelsVisible / height;
+
+        if (percentVisible > 0.8) {
+          return true;
+        }
+      }
+
+      return false;
+    };
+    function getAllProductsOnPage() {
+      const allAnchorTags = document.getElementsByTagName("a");
+      const productUrlRegex = new RegExp(
+        `^(?!\/\/cdn)[-\.:\/,a-z,A-Z,0-9]*\/products\/`
+      );
+      const allAnchorTagsWithProducts = Array.prototype.slice
+        .call(allAnchorTags)
+        .filter((anchorTag) => {
+          return productUrlRegex.test(anchorTag.href);
+        });
+      let allProductsOnPageWithImg = allAnchorTagsWithProducts.filter(
+        (anchorTag) => {
+          anchorTag.querySelector("img");
+        }
+      );
+      if (allProductsOnPageWithImg.length === 0) {
+        allProductsOnPageWithImg = allAnchorTagsWithProducts.filter(
+          (anchorTag) => {
+            const parentNode = anchorTag.parentNode.parentNode.parentNode;
+            return (
+              parentNode.nextElementSibling?.querySelector("img") ||
+              parentNode.previousElementSibling?.querySelector("img")
+            );
+          }
+        );
+      }
```

Comment:
Can you explain what are we looking for here?

Reply:
We are fetching all the anchor tags from a page. After that, we check whether any `img` element surrounds that anchor tag. The logic of checking three levels is taken from littledata, it might be the generic structure for Shopify.